cpp/src/arrow/array/array_test.cc

```diff
@@ -3348,4 +3358,54 @@ TEST(TestSwapEndianArrayData, InvalidLength) {
   }
 }
 
+template <typename T>
+class TestPrimitiveArray : public ::testing::Test {
+ public:
+  void SetUp() { pool_ = default_memory_pool(); }
+
+ protected:
+  MemoryPool* pool_;
+};
+
+TYPED_TEST_SUITE(TestPrimitiveArray, Primitives);
+
+TYPED_TEST(TestPrimitiveArray, IndexOperator) {
+  using ArrayBuilderPointer = std::shared_ptr<arrow::ArrayBuilder>;
+  using CurrentBuilderType = typename TypeParam::BuilderType;
+  using CurrentArrayType = typename TypeParam::ArrayType;
+  using T = typename TypeParam::T;
+
+  using CType = typename std::conditional<std::is_integral<T>::value ||
+                                              std::is_floating_point<T>::value,
+                                          T, typename TypeParam::ConversionType>::type;
+
+  const std::vector<bool> valids = {true, true, true, false, true, true};
+  const std::vector<CType> values(
+      6);  // default/garbage values, not important for this test
```

Comment:
@pitrou It had to be two types of vectors (`{{1,1}, {1,1}}` and `{1, 1}`, Hence removed and left with default values as actual value is not important for checking if `[]` is returning original value. Please let me know your thoughts.

Reply:
Hmm, no, we should never use undefined values in tests. Also, in this case you would get zeros, which isn't very interesting.

There could be several ways to achieve this, for example you can try using `RandomArrayGenerator::ArrayOf`.

Reply:
That is a good lead, I am still getting used to different approaches on how to create arrays in arrow. Thanks